handbook/engineering/observability/monitoring_pillars.md

```diff
@@ -63,28 +35,61 @@ Before the constraints we impose will start to make sense, you need to understan
 - It has been impossible/painful for site admins to define meaningful alerting for Sourcegraph. Many admins would fall back to poor external checks like "does a search query work and is it fast?"
 - It has been impossible for us to debug customer's instances through monitoring, and often we would just write one-off Prometheus queries for customers to run.
 
-### The linter of monitoring
-
 We can and should do better with monitoring than we have in the past, so we introduce strict restrictions on how we develop monitoring. This helps keep the entire engineering team aligned with what our customers and users need from us -- even if monitoring is often an afterthought or the last thing an engineer working on some feature adds.
 
 You can think of the restrictions we impose as a linter for preventing the issues described above. And, in many cases, our tooling does already prevent these issues by imposing these restrictions!
 
-## Frequently asked questions
+## The Sourcegraph monitoring pillars
+
+Monitoring tooling at Sourcegraph is developed to encourage the following guidelines:
+
+- [**Dashboards should only be created with the monitoring generator**](#dashboards-should-only-be-created-with-the-monitoring-generator)
+- [**Dashboards should describe a single service**](#dashboards-should-describe-a-single-service)
+- [**Graphs should have associated alerts**](#graphs-should-have-associated-alerts)
+- [**Graphs should have less than 5 cardinality**](#graphs-should-have-less-than-5-cardinality)
+- [**Less useful graphs should be hidden by default**](#less-useful-graphs-should-be-hidden-by-default)
+
+Before trying to circumvent the guidelines enforced by our monitoring tooling, please keep in mind that these pillars are defined to help us achieve our [goals](#long-term-vision) and alleviate [pain points we have encountered in the past](#history). Each of the above pillars are documented with detailed explanations, recommendations for what you can do, and exceptions where applicable.
+
+To make significant changes to the monitoring tooling, please reach out to the [Distribution team](../distribution/index.md) or open a PR to this page!
+
+### Dashboards should only be created with the monitoring generator
+
+Creating dashboards outside of the monitoring generator, such as with the Grafana WYSIWYG editor, is discouraged. All metrics should be presented in the most plain, mundane, non-fanciful way that every site admin can understand.
```

Comment:
Also disagree with this. The terms "mundane" or "non-fanciful" are completely subjective. Using whatever Grafana features to improve how you understand and navigate your graphs should be encouraged, not the other way around.

I want to try and expose the Grafana SDK in a much more direct manner in the generator.

Reply:
Maybe it can be rephrased, but I do agree that we need to provide customers with easy to understand the health of the services, keeping in mind that not all customers are as technical and almost all of them do not have our context and understand of the systems.

Reply:
How about this approach instead:

```suggestion
Creating dashboards outside of the monitoring generator, such as with the Grafana WYSIWYG editor, is discouraged. All metrics should be presented in a set of uniform, standard ways that every site admin can understand.
```

From what I've seen, I feel like most dashboards fall into a few categories (ones that work well currently, ones that don't mostly high-cardinality graphs, that could benefit from things like labels on the right). We can create a set of formatting options that can be shared amongst everyone trying to create particular types of graphs

> keeping in mind that not all customers are as technical and almost all of them do not have our context and understand of the systems.

Lots of colours and elaborate formatting might be nice, but I think they definitely hurt the goal of making dashboards easy to interpret at a glance - _some_ expectation of how a dashboard will look going from service to service is extremely valuable IMO, especially for the uninitiated

Reply:
>  some expectation of how a dashboard will look going from service to service is extremely valuable IMO, especially for the uninitiated

Yes, it is, but those standards and reusable patterns should evolve over time, and that can only happen if we allow experimentation. For instance, we don't consistently show any implementation of USE or RED for [Golden Signals](https://landing.google.com/sre/sre-book/chapters/monitoring-distributed-systems/#xref_monitoring_golden-signals), I think that's where we should start.

Reply:
My suggestion, a bit simpler.
```suggestion
Creating dashboards outside of the monitoring generator, such as with the Grafana WYSIWYG editor, is discouraged. All metrics should be presented in a way that every site admin can understand.
```

> Yes, it is, but those standards and reusable patterns should evolve over time, and that can only happen if we allow experimentation. For instance, we don't consistently show any implementation of USE or RED for Golden Signals, I think that's where we should start.

I dont think if this is at odds with this or prevents us from creating USE and RED metrics and dashboards for those.

Reply:
It's a really high bar. Not all site admins will understand Sourcegraph's architecture well enough that every single panel is obvious to them. What if a site-admin doesn't know that the frontend sends git commands to git server which executes them and sends back the raw output? Does that mean we shouldn't have a graph with rate of git commands over time per git server pod? Where do we draw the line?

Reply:
> Does that mean we shouldn't have a graph with rate of git commands over time per git server pod

I don't think this pillar is at odds with a graph like this (you might be referring to the pillar about cardinality, where we can easily say "use your best judgement, if this information only makes sense on a per-instance level"). Using the generator simply says the _presentation_ must be minimal and standard. From the rationale for this pillar:

> [not using the generator makes it] Extremely easy to introduce dashboards that have inconsistency errors through mistakes that are easy to make (e.g. not using UTC time on dashboards, the preset timeframe being inconsistent, it not self-describing to admins what it does, etc.)

Also, big hunks of JSON is hard to review and make changes to:

> [not using the generator makes it] Difficult for others to modify and review without introducing regressions

This also goes back to my previous comment:

> From what I've seen, I feel like most dashboards fall into a few categories [...] This means improvements can be more easily shared, and make future graphs easier to create

For example, you mention that the provisioning graphs look _awful_, which is true! Anything we do to make the provisioning graphs look _less_ awful, however, can easily be applied to a wide range of other graphs (namely, anything graphed per-instance), which I think is what this pillar is getting at - the generator makes it easy to make minimal, standard, consistent graphs that makes it simpler to understand.

> Yes, it is, but those standards and reusable patterns should evolve over time, and that can only happen if we allow experimentation

We do allow experimentation, since anyone can propose changes to the generator.

Could we better support this by renaming `Internal` dashboards to `Experimental`, with the expectation that they be graduated to generator dashboards?

* Remove `internal` entirely from `sourcegraph/grafana`
* Create a workflow to provision `experimental` dashboards directly through `ConfigMap` in `sourcegraph.com`

This would mean that teams can feel more free to iterate on experimental and dot-com-specific  dashboards without worrying about how site admins interpret it

Reply:
> This would mean that teams can feel more free to iterate on experimental and dot-com-specific dashboards without worrying about how site admins interpret it

+1, but I don't know if provisioning these through a ConfigMap is better than via the Docker image. Also, I think internal is fine. What think prevented experimentation is 1) lack of access to more Grafana features and 2) an overly restrictive set of "pillars" that felt quite immovable for others (i.e. didn't invite collaboration).

Reply:
> I don't know if provisioning these through a ConfigMap is better than via the Docker image

Removes the worry of shipping potentially confusing dashboards to customers, and adds more impetus to work towards integrating them into the standard generator dashboards to get them shipped

> an overly restrictive set of "pillars" that felt quite immovable for others (i.e. didn't invite collaboration).

Hopefully the new more suggestion-y wording on the pillars improve this!

Reply:
SGTM

Reply:
> > some expectation of how a dashboard will look going from service to service is extremely valuable IMO, especially for the uninitiated
> 
> Yes, it is, but those standards and reusable patterns should evolve over time, and that can only happen if we allow experimentation. For instance, we don't consistently show any implementation of USE or RED for [Golden Signals](https://landing.google.com/sre/sre-book/chapters/monitoring-distributed-systems/#xref_monitoring_golden-signals), I think that's where we should start.

I think the goal that we want dashboards to be consistent is valuable, so I'm going to go with my version of this pillar, unless anyone objects: https://github.com/sourcegraph/about/pull/2000#discussion_r523055237

I will see if I can squeeze in more wording about how these pillars and our tooling are open to improvement and iteration, but most of that stuff might have to wait for some implementation to catch up: https://github.com/orgs/sourcegraph/projects/112